cli/command/image/build_buildkit.go

```diff
@@ -0,0 +1,311 @@
+package image
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/containerd/console"
+	"github.com/docker/cli/cli"
+	"github.com/docker/cli/cli/command"
+	"github.com/docker/cli/cli/command/image/build"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/pkg/jsonmessage"
+	"github.com/docker/docker/pkg/stringid"
+	"github.com/docker/docker/pkg/urlutil"
+	controlapi "github.com/moby/buildkit/api/services/control"
+	"github.com/moby/buildkit/client"
+	"github.com/moby/buildkit/session/auth/authprovider"
+	"github.com/moby/buildkit/session/filesync"
+	"github.com/moby/buildkit/util/appcontext"
+	"github.com/moby/buildkit/util/progress/progressui"
+	"github.com/pkg/errors"
+	"github.com/tonistiigi/fsutil"
+	"golang.org/x/sync/errgroup"
+)
+
+const uploadRequestRemote = "upload-request"
+
+var errDockerfileConflict = errors.New("ambiguous Dockerfile source: both stdin and flag correspond to Dockerfiles")
+
+//nolint: gocyclo
+func runBuildBuildKit(dockerCli command.Cli, options buildOptions) error {
+	ctx := appcontext.Context()
+
+	s, err := trySession(dockerCli, options.context)
+	if err != nil {
+		return err
+	}
+	if s == nil {
+		return errors.Errorf("buildkit not supported by daemon")
+	}
+
+	buildID := stringid.GenerateRandomID()
+
+	var (
+		remote           string
+		body             io.Reader
+		dockerfileName   = options.dockerfileName
+		dockerfileReader io.ReadCloser
+		dockerfileDir    string
+		contextDir       string
+	)
+
+	switch {
+	case options.contextFromStdin():
+		if options.dockerfileFromStdin() {
+			return errStdinConflict
+		}
+		rc, isArchive, err := build.DetectArchiveReader(os.Stdin)
+		if err != nil {
+			return err
+		}
+		if isArchive {
+			body = rc
+			remote = uploadRequestRemote
+		} else {
+			if options.dockerfileName != "" {
+				// This is new behavior, but @tiborvass believes it's correct.
```

Comment:
This looks like a discussion for the commit message rather than a code comment.

Reply:
@ijc addressed in commit message.